source/constructs/api/data_source/service.py

```diff
@@ -1193,41 +1212,221 @@ def before_delete_rds_connection(account: str, region: str, instance: str):
+    # else:
+    #     delete_glue_connection(account, region, crawler_name, glue_database_name, glue_connection_name)
+    try:
+        # PENDING｜ACTIVE｜ERROR with message
+        crud.set_s3_bucket_source_glue_state(account, region, bucket, ConnectionState.PENDING.value)
+        iam_role_name = crud.get_iam_role(account)
+        assumed_role = sts.assume_role(
+            RoleArn=f"{iam_role_name}",
+            RoleSessionName="glue-s3-connection"
+        )
+        credentials = assumed_role['Credentials']
+        s3_targets = build_s3_targets(bucket, credentials, region, True)
+        logger.info("sync_s3_connection rebuild s3 targets:")
+        logger.info(s3_targets)
+        glue = boto3.client('glue',
+                            aws_access_key_id=credentials['AccessKeyId'],
+                            aws_secret_access_key=credentials['SecretAccessKey'],
+                            aws_session_token=credentials['SessionToken'],
+                            region_name=region
+                            )
+        """ :type : pyboto3.glue """
+
+        try:
+            glue.get_database(Name=glue_database_name)
+        except Exception as e:
+            logger.info("sync_s3_connection glue get_database error, and create database")
+            logger.info(e)
+            response = glue.create_database(DatabaseInput={'Name': glue_database_name})
+            logger.info(response)
+        # wait for database creation, several seconds
+        lakeformation = boto3.client('lakeformation',
+                                     aws_access_key_id=credentials['AccessKeyId'],
+                                     aws_secret_access_key=credentials['SecretAccessKey'],
+                                     aws_session_token=credentials['SessionToken'],
+                                     region_name=region)
+        """ :type : pyboto3.lakeformation """
+        crawler_role_arn = __gen_role_arn(account_id=account,
+                                          region=region,
+                                          role_name='GlueDetectionJobRole')
+
+        # retry for grant permissions
+        num_retries = GRANT_PERMISSIONS_RETRIES
+        while num_retries > 0:
+            try:
+                response = lakeformation.grant_permissions(
+                    Principal={
+                        'DataLakePrincipalIdentifier': f"{crawler_role_arn}"
+                    },
+                    Resource={
+                        'Database': {
+                            'Name': glue_database_name
+                        }
+                    },
+                    Permissions=['ALL'],
+                    PermissionsWithGrantOption=['ALL']
+                )
+            except Exception as e:
+                sleep(SLEEP_MIN_TIME)
+                num_retries -= 1
+            else:
+                break
+        else:
+            raise Exception('UNCONNECTED')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.